keps/976-plain-pods/README.md

```diff
@@ -0,0 +1,695 @@
+Beta, allows an external controller to prevent kube-scheduler from scheduling Pods. Kueue can make
+use of this API to implement queuing semantics for Pods.
+
+### Goals
+
+- Support queueing of individual Pods.
+- Support queueing of groups of Pods of fixed size, identified by a common label or annotation.
+- Opt-in or opt-out Pods from specific namespaces from queuing.
+
+### Non-Goals
+
+- Support for [dynamic reclaiming quota](https://github.com/kubernetes-sigs/kueue/issues/78)
+
+  This feature is incompatible with supporting Pod replacements without knowing the behavior of a
+  parent controller for the Pods.
+
+- Support for [partial-admission](https://github.com/kubernetes-sigs/kueue/issues/420).
+
+  Since all pods are already created, an implementation of partial admission would imply the
+  deletion of some pods. It is not clear if this matches users expectations, as opposed to support
+  for elastic groups.
+
+- Support elastic groups of Pods, where the number of Pods changes after the job started.
+
+  While these jobs are one of the motivations for this KEP, the current proposal doesn't support
+  them. These jobs can be addressed in follow up KEPs.
+
+- Support for advanced Pod retry policies
+
+  Kueue shouldn't re-implement core functionalities that are already available in the Job API.
+
+- Tracking usage of Pods that were not queued through Kueue.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### User Stories (Optional)
+
+#### Story 1
+
+As an platform developer, I can queue plain Pods. I just add a queue name to the Pods through a
+label.
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: foo
+  namespace: pod-namespace
+  labels:
+    kueue.x-k8s.io/queue-name: user-queue
+spec:
+  containers:
+  - name: job
+    image: hello-world
+    resources:
+      requests:
+        cpu: 1m
+```
+
+#### Story 2
+
+As a platform developer, I can queue groups of Pods that share the same shape (Pod specs).
+In addition to the queue name, I can specify how many Pods belong to the group.
+
+<<[UNRESOLVED usability of labels and annotations ]>>
+Could the split among labels and annotations cause user mistakes?
+Note that integers cannot be used as label values.
+<<[/UNRESOLVED]>>
+
+The pods of a job with a single spec (similar to Pods in an Indexed Job), look like follows:
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: pod-index-0
+  namespace: pod-namespace
+  labels:
+    kueue.x-k8s.io/queue-name: user-queue
+    kueue.x-k8s.io/pod-group-name: pod-group
+  annotations:
+    kueue.x-k8s.io/pod-group-count: "10"
```

Comment:
It would be great to make this optional and deduct the max by waiting for e.g. 5 minutes. Where that 5 minute timer is also adjustable through a label.

Reply:
I think it might be better to work on elastic jobs, instead of that. For now, I hope that having a single parameter for total number of pods is enough.